netlogo-gui/src/main/editor/EditorArea.scala

```diff
@@ -0,0 +1,435 @@
+      // start - ST 12/20/04
+      setCaretPosition(getText().length)
+      moveCaretPosition(0)
+    }
+    Actions.setEnabled(true)
+    UndoManager.setCurrentManager(undoManager)
+  }
+
+  def focusLost(fe: java.awt.event.FocusEvent): Unit = {
+    // On Windows (and perhaps Linux? not sure), putting
+    // the focus elsewhere leaves the text selected in the
+    // now-unfocused field.  This causes the text to be drawn
+    // in different colors even though the selection isn't
+    // visible.  I suppose we could make HighlightView smarter
+    // about that, but instead let's just force the Mac-like
+    // behavior and be done with it for now - ST 11/3/03
+    if (enableFocusTraversalKeys)
+      select(0, 0)
+    mouseEvent = fe.isTemporary
+    bracketMatcher.focusLost(this)
+    colorizer.reset()
+    if (!fe.isTemporary) {
+      Actions.setEnabled(false)
+      UndoManager.setCurrentManager(null)
+    }
+  }
+
+  // this is used for quick help, when QH is triggered
+  // by the context menu we want to look up the word under
+  // the mouse pointer without moving the cursor ev 7/3/07
+  private var mousePos: Int = 0
+
+  def getMousePos: Int = mousePos
+
+  override def processMouseEvent(me: java.awt.event.MouseEvent): Unit = {
+    if (me.getID == java.awt.event.MouseEvent.MOUSE_PRESSED) {
+      mouseEvent = true
+    }
+    if (me.isPopupTrigger() && ! contextMenu.isShowing()) {
+      mousePos = caret.getMousePosition(me)
+      doPopup(me)
+    } else
+      super.processMouseEvent(me)
+  }
+
+  private class EditorContextMenu(colorizer: Colorizer, i18n: String => String) extends JPopupMenu {
+
+    val copyItem  = new JMenuItem(Actions.COPY_ACTION)
+    val cutItem   = new JMenuItem(Actions.CUT_ACTION)
+    val pasteItem = new JMenuItem(Actions.PASTE_ACTION)
+
+    locally {
+      add(copyItem)
+      Actions.COPY_ACTION.putValue(Action.NAME, i18n.apply("menu.edit.copy"))
+      add(cutItem)
+      Actions.CUT_ACTION.putValue(Action.NAME, i18n.apply("menu.edit.cut"))
+      add(pasteItem)
+      Actions.PASTE_ACTION.putValue(Action.NAME, i18n.apply("menu.edit.paste"))
+      addSeparator()
+      add(new JMenuItem(Actions.mouseQuickHelpAction(colorizer, i18n)))
+      for(item <- menuItems) {
+        item.putValue("editor", EditorArea.this)
+        add(new JMenuItem(item))
+      }
+    }
+
+    override def show(invoker: Component, x: Int, y: Int): Unit = {
+      val text = EditorArea.this.getSelectedText
+      val isTextSelected = Option(text).exists(_.length > 0)
+      copyItem.setEnabled(isTextSelected)
+      cutItem.setEnabled(isTextSelected)
+      pasteItem.setEnabled(
+        Toolkit.getDefaultToolkit.getSystemClipboard.isDataFlavorAvailable(DataFlavor.stringFlavor))
+      val point = new Point(invoker.getLocationOnScreen)
+      point.translate(x, y)
+      for(item <- menuItems){
+        item.putValue("cursorLocation", mousePos)
+        item.putValue("popupLocation", point)
```

Comment:
I like this a lot. It's a nice method of providing data to actions that need it without coupling the editor to the individual actions.